src/installer/iohandler.py

```diff
@@ -0,0 +1,100 @@
+"""Handling wheel installation I/O-related functionality."""
+
+import hashlib
+import os
+from io import RawIOBase
+
+from installer._compat.typing import TYPE_CHECKING
+
+from .records import Hash, Record
+
+if TYPE_CHECKING:
+    from ._compat.typing import FSPath
+    from typing import Dict, Tuple, Optional, Callable
+
+_BLOCK_SIZE = 4096
+_SCHEME_KEYS = ["platlib", "purelib"]
+
+
+def basic_copy_handler(
+    source,  # type: RawIOBase
+    dest,  # type: FSPath
+    hash_algorithm="sha256",  # type: Optional[str]
+):
+    # type: (...) -> Tuple[int, Hash]
+    """Copy a buffer and compute its hash and size.
+
+    Copies the source buffer to the specified destination. It also computes the hash
+    of the file to avoid traversing the source twice.
+
+    :param source: buffer holding the source data
+    :param dest: destination path
+    :param hash_algorithm: hashing algorithm
+
+    :return: size and ``Hash`` tuple of the copied buffer
+    """
+    assert hash_algorithm is not None  # to please mypy
+    hasher = hashlib.new(hash_algorithm)
+    size = 0
+    with open(dest, "wb") as dest_fd:
+        while True:
+            buf = source.read(_BLOCK_SIZE)
+            if not buf:
+                break
+            hasher.update(buf)
+            dest_fd.write(buf)
+            size += len(buf)
+
+    hash_ = Hash.parse("sha256={}".format(hasher.hexdigest()))
+    return size, hash_
+
+
+class IOHandler(object):
+    """Implement I/O operations for installers."""
+
+    def __init__(
+        self,
+        scheme_path_map,  # type: Dict[str, FSPath]
+        hash_algorithm="sha256",  # type: Optional[str]
+        copy_handler=basic_copy_handler,  # type: Callable[[RawIOBase, FSPath, Optional[str]], Tuple[int, Hash]]
```

Comment:
I'd make this type aliased in the `if TYPE_CHECKING`, via `CopyHandler = ...` assignment there.